### PR TITLE
Add category page (close #1378)

### LINKED ISSUE #1378: Show search results for categories on /categories/ pages
Allow `/extensions/alerts-updates/` URLs to show all add-ons in a category.

### PR DESCRIPTION
**Pull request by tofumatt**
Add the actual category URL (matches addons-server) including pagination.

Screenshot:

nov-21-2016 12-21-02

### CODE REVIEW

File: tests/client/amo/containers/TestCategoryPage.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import deepcopy from 'deepcopy';
+
+import { mapStateToProps } from 'amo/containers/CategoryPage';
+
+
+describe('CategoryPage.mapStateToProps()', () => {
+  const state = {
```
Comment (kumar303) [inline tests/client/amo/containers/TestCategoryPage.js]: I think it would be less fragile to create this state by dispatching real actions. It should be just as easy to do. Example: https://github.com/mozilla/addons-frontend/issues/1371

File: tests/client/amo/containers/TestCategoryPage.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      },
+    });
+
+    assert.deepEqual(props, {
+      hasSearchParams: true,
+      pathname: '/themes/ad-block/',
+      queryParams: { page: 1 },
+    });
+  });
+});
```
Comment (kumar303) [inline tests/client/amo/containers/TestCategoryPage.js]: This appears to be missing a test for the case when `state.search` is empty (i.e. `{}`).

File: tests/client/core/test_searchUtils.js [hunk 3]
```diff
@@ -16,3 +21,77 @@ describe('searchUtils mapStateToProps()', () => {
     assert.deepEqual(props, { hasSearchParams: false });
   });
 });
+
+describe('searchUtils loadCategoryResultsIfNeeded()', () => {
+  const state = {
```
Comment (kumar303) [inline tests/client/core/test_searchUtils.js]: I'd also like to this state built by dispatching real actions, if possible

File: tests/client/core/test_searchUtils.js [hunk 4]
```diff
@@ -16,3 +21,77 @@ describe('searchUtils mapStateToProps()', () => {
...
+      getState: () => stateWithPage,
+    };
+    const location = { query: { page } };
+    const params = {
+      addonType: 'extension',
+      application: 'firefox',
+      slug: 'ad-block',
+    };
+    assert.strictEqual(
+      loadCategoryResultsIfNeeded({ store, location, params }), true);
```
Comment (kumar303) [inline tests/client/core/test_searchUtils.js]: I think this test would be better if it verified that `mockApi` is never called, which is a stronger indicator that data is not fetched from the API.

File: src/core/searchUtils.js [hunk 5]
```diff
@@ -92,6 +92,22 @@ export function loadSearchResultsIfNeeded({ store: { dispatch, getState }, locat
...
     clientApp: state.api.clientApp,
   });
+
+  if (!isLoaded({ state: state.search, page, filters })) {
+    return performSearch({ dispatch, page, api: state.api, auth: state.auth, filters });
+  }
+  return true;
+}
+
+export function loadCategoryResultsIfNeeded({ store: { dispatch, getState }, location, params }) {
```
Comment (kumar303) [inline src/core/searchUtils.js]: This could be shortened slightly to `loadByCategoryIfNeeded`

File: src/core/searchUtils.js [hunk 6]
```diff
@@ -92,6 +92,22 @@ export function loadSearchResultsIfNeeded({ store: { dispatch, getState }, locat
...
+  const page = parsePage(location.query.page);
+  const state = getState();
+  const filters = {
+    addonType: params.addonType,
+    category: params.slug,
+    clientApp: params.application,
+  };
+
   if (!isLoaded({ state: state.search, page, filters })) {
```
Comment (kumar303) [inline src/core/searchUtils.js]: Why do the search results need to be cached? I think it should make an API request every time. Otherwise, the single page app will go stale -- the only way a user can obtain newly added results is to refresh the page. The browser will already cache the API call as needed so I don't really see the value in adding a new caching layer on top of that.

File: src/core/searchUtils.js [hunk 7]
```diff
@@ -92,6 +92,22 @@ export function loadSearchResultsIfNeeded({ store: { dispatch, getState }, locat
     ...location.query,
     clientApp: state.api.clientApp,
   });
+
+  if (!isLoaded({ state: state.search, page, filters })) {
+    return performSearch({ dispatch, page, api: state.api, auth: state.auth, filters });
+  }
+  return true;
```
Comment (tofumatt) [inline src/core/searchUtils.js]: I agree actually, as it makes testing weird too. That said:

it-was-like-that-when-i-got-here

I've filed an issue for it (#1403), as changing this across the app would be out-of-scope for the PR.

File: src/core/searchUtils.js (hunk 6 above)
Comment (tofumatt) [inline src/core/searchUtils.js]: Good point. Again, it was like that when I got here; the search page currently caches results. When flipping through paginated results it makes going back and forth between pages quick.

I've filed #1404, but I'll remove it from this bit of code because why not and I agree; we can always add it back.

File: src/amo/containers/CategoryPage.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+import SearchPage from 'amo/components/SearchPage';
+import {
+  filtersMatch,
+  loadCategoryResultsIfNeeded,
+  parsePage,
+} from 'core/searchUtils';
+
+
+export function mapStateToProps(state, ownProps) {
+  const filters = {
```
Comment (tofumatt) [inline src/amo/containers/CategoryPage.js]: But these filters are declared here to compare them with state and make sure everything matches. They're passed to `if (filtersMatch([...])) {`, so moving them into there would mean duplication.

They're only passed to the state if they match.

File: tests/client/amo/containers/TestCategoryPage.js (hunk 2 above)
Comment (tofumatt) [inline tests/client/amo/containers/TestCategoryPage.js]: I don't think that should ever happen, as it's set by `searchStart`. Am I missing a path to an empty search state?

**Comment by tofumatt** [comment]
Okay, I think I fixed all the issues (mostly regarding state testing) in the PR, and created issues for the things that were slightly out-of-scope. Ready for another r?

**Comment by tofumatt** [comment]
Odd: https://travis-ci.org/mozilla/addons-frontend/builds/177831776 is failing but passing locally, and the error is quite strange.

**Comment by mstriemer** [comment]
Try rebasing with master. That's what I was running into yesterday and switching back to npm seemed to fix it.

**Comment by tofumatt** [comment]
Ah right, cool. Figured this was related, cheers.

File: src/amo/containers/CategoryPage.js (hunk 8 above)
Comment (kumar303) [inline src/amo/containers/CategoryPage.js]: Ah, ok, I see on a second look that all of these do need to be defined outside of the if block.

File: src/core/searchUtils.js [hunk 9]
```diff
@@ -92,8 +92,28 @@ export function loadSearchResultsIfNeeded({ store: { dispatch, getState }, locat
...
     clientApp: state.api.clientApp,
   });
+
   if (!isLoaded({ state: state.search, page, filters })) {
     return performSearch({ dispatch, page, api: state.api, auth: state.auth, filters });
   }
   return true;
 }
+
+export function loadByCategoryIfNeeded(
```
Comment (kumar303) [inline src/core/searchUtils.js]: This name no longer makes sense :) I think `loadByCategory` would be fine.

File: tests/client/amo/containers/TestCategoryPage.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+      queryParams: { page: 1 },
+      results: [],
+    });
+  });
+
+  it('does not pass search state if the filters and state do not match', () => {
+    const store = createStore();
+    store.dispatch(searchStart({ filters }));
+    const mismatchedState = store.getState();
+    mismatchedState.search.filters.clientApp = 'nothing';
```
Comment (kumar303) [inline tests/client/amo/containers/TestCategoryPage.js]: Using dispatch (like above) is the way to do it IMO. By using dispatch you are not coupling this code with the raw state representation. Here would be a way to rewrite it:
````javascript
store.dispatch(searchStart({ filters: { ...filters, clientApp: 'nothing' } }));
const mismatchedState = store.getState();
````

File: tests/client/amo/containers/TestCategoryPage.js (hunk 2 above)
Comment (kumar303) [inline tests/client/amo/containers/TestCategoryPage.js]: The component may render with `state.search = {}` for a split second if the API response is slow but, no worries, we could always add the test later if a bug appears with handling empty state.

File: tests/client/core/test_searchUtils.js [hunk 11]
```diff
@@ -16,3 +20,38 @@ describe('searchUtils mapStateToProps()', () => {
     assert.deepEqual(props, { hasSearchParams: false });
   });
 });
+
+describe('searchUtils loadByCategoryIfNeeded()', () => {
```
Comment (kumar303) [inline tests/client/core/test_searchUtils.js]: `loadByCategoryIfNeeded` would need to be renamed too

File: src/core/searchUtils.js (hunk 9 above)
Comment (mstriemer) [inline src/core/searchUtils.js]: Is this making another API request on the client after JS loading on initial render? I think that's why we've normally included the caching.

File: tests/client/amo/containers/TestCategoryPage.js (hunk 10 above)
Comment (tofumatt) [inline tests/client/amo/containers/TestCategoryPage.js]: Oops, good point. Will do.
